runelite-client/src/main/java/net/runelite/client/plugins/camera/CameraPlugin.java

```diff
@@ -117,36 +127,83 @@ protected void shutDown()
 		controlDown = false;
 	}
 
+	@Subscribe
+	public void onGameTick(GameTick gameTick)
+	{
+		if (currentLeftClickOption.equals(defaultLeftClickOption) || client.getTickCount() < lastUpdatedTick + RESET_DIRECTION_TIMEOUT)
+		{
+			return;
+		}
+		currentLeftClickOption = defaultLeftClickOption;
+	}
+
 	@Subscribe
 	public void onMenuEntryAdded(MenuEntryAdded menuEntryAdded)
 	{
-		if (menuEntryAdded.getType() == MenuAction.CC_OP.getId() && menuEntryAdded.getOption().equals(LOOK_NORTH) && config.compassLook())
+		if (menuEntryAdded.getType() == MenuAction.CC_OP.getId())
 		{
-			MenuEntry[] menuEntries = client.getMenuEntries();
-			int len = menuEntries.length;
-			MenuEntry north = menuEntries[len - 1];
+			if (ArrayUtils.indexOf(options, menuEntryAdded.getOption()) < 0)
+			{
+				return;
+			}
 
-			menuEntries = Arrays.copyOf(menuEntries, len + 3);
+			final int count = options.length;
+			MenuEntry[] menuEntries = client.getMenuEntries();
+			int currentOptionIndex = ArrayUtils.indexOf(options, currentLeftClickOption);
 
-			// The handling for these entries is done in ToplevelCompassOp.rs2asm
-			menuEntries[--len] = createCameraLookEntry(menuEntryAdded, 4, LOOK_WEST);
-			menuEntries[++len] = createCameraLookEntry(menuEntryAdded, 3, LOOK_EAST);
-			menuEntries[++len] = createCameraLookEntry(menuEntryAdded, 2, LOOK_SOUTH);
-			menuEntries[++len] = north;
+			if (config.compassLook())
+			{
+				menuEntries = Arrays.copyOf(menuEntries, menuEntries.length + 3);
+
+				// fill the menu from top to bottom
+				// but iterate through options cyclically from 0 -> 3 -> 0
+				int entryIndex = menuEntries.length - 1;
+
+				// The handling for these entries is done in ToplevelCompassOp.rs2asm`
+				menuEntries[entryIndex] = createCameraLookEntry(menuEntryAdded, currentOptionIndex + 1, options[currentOptionIndex % count]);
+				entryIndex -= 1;
+				currentOptionIndex += 1;
```

Comment:
Lines 164-165 could be removed with something like `menuEntries[entryIndex--]` and `currentOptionIndex++` but I found it hard to follow when written that way